website/blog/2023-4-10-incrementals-pt1.md

```diff
@@ -0,0 +1,140 @@
+Corso implementation, so we thought we’d share them here."
+authors: amartinez
+tags: [corso, microsoft 365, backups]
+date: 2023-04-09
+image: ./images/incremental-encoder.jpg
+---
+<!-- vale Vale.Spelling = NO -->
+
+![By © Raimond Spekking / CC BY-SA 4.0 (via Wikimedia Commons), CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=75914553](./images/incremental-encoder.jpg)
+<!-- vale Vale.Spelling = YES -->
+
+Full Microsoft 365 backups can take a long time, especially since Microsoft
+throttles the number of requests an application can make in a given window of
+time. Recent additions to Corso have reduced the duration of backups after the
+first backup by taking advantage of Microsoft’s delta query API. Doing so allows
+Corso to retrieve only the changes to the user’s data since the last backup
+instead of having to retrieve all items with the Graph API. However,
+implementing backups in this manner required us to play a few tricks with the
+Corso implementation, so we thought we’d share them here.
+
+<!-- truncate -->
+
+## Background
+
+Before we dive into the details of how incremental backups work, it’s useful to
+have some knowledge of how delta queries work in the Microsoft Graph API and how
+data is laid out in Corso backups.
+
+### Microsoft delta queries
+
+Microsoft provides a delta query API that allows developers to get only the
+changes to the endpoint since the last query was made. The API represents the
+idea of the “last query” with an opaque token that's returned when the set of
+items is done being listed. For example, if a developer wants to get a delta
+token for a specific email folder, the developer would first list all the items
+in the folder using the delta endpoint. On the final page of item results from
+the endpoint, the Graph API would return a token that could be used to retrieve
+future updates.
+
+All returned tokens represent a point in time and are independent from each
+other. This means that getting token a1 at time t1, making some changes, and
+then getting another token a2 at time t2 would give distinct tokens. Requesting
+the changes from token a1 would always give the changes made after time t1
+including those after time t2. Requesting changes from token a2 would give only
+the changes made after time t2. Tokens eventually expire though, so waiting a
+long time between backups (e.x. weeks) may cause all items to be enumerated
+again.
+
+## Corso full backups, incremental backups, and backup layout
+
+Before we get into the nuts and bolts of how Corso uses the Microsoft delta
+query API, it’s important to first define what’s in a backup and the terminology
+we’ll be using throughout this post.
+
+### Backup layout
+
+Internally, a single Corso backup consists of three main parts: a kopia manifest
+that Corso uses as the root object of the backup (BackupModel), a kopia snapshot
+of indexing information for Corso, and a kopia snapshot of the item data in the
```

Comment:
The name is is a bit awkward and especially for "kopia snapshot of ..." Maybe make it "backup index" and "the data backup"